File: Java/src/Jump Game II.java
public class Solution {
    public int jump(int[] A) {
        if (A.length <= 1) return 0;
        int maxReach = A[0];
        int edge = 0;
        int ret = 0;
        
        for(int i = 0 ; i < A.length; i++){
            if (i > edge){
                ret++;
                edge = maxReach;
                if (edge > A.length - 1) return ret;
            }
            maxReach = Math.max(maxReach, A[i] + i);
            if (maxReach < i) return -1;
        }
        return ret;
    }
}

File: Python/Jump Game II.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def jump(self, A):
        if len(A) == 1:
            return 0
        maxReach = A[0]
        edge = 0
        ret = 0

        for i in range(len(A)):
            if i > edge:
                ret += 1
                edge = maxReach

                if edge > len(A) - 1:
                    return ret
            maxReach = max(maxReach, A[i] + i)
            if maxReach < i: return -1
            print i, edge, maxReach

        return ret
    
s = Solution()
print s.jump([2,3,1,1,4])